# Analysis of a similarity network of Tc-mar transposon copies

## Abstract
A network with nodes as Mariner-type transposon consensus sequences from *C. brenneri* genomes and edges representing sequences that passed a similarity threshold was acquired from Scheinder et al. (2021). Various properties of the network were analyzed, including closeness centrality, degree distribution, edge_betweeness, average neighbor degree, and clustering coefficient. These metrics allow for the quantitative analysis of the evolutionary dynamics of lineages, and analysis in the manner has been performed here. 

## Motivation
It has recently been proposed that biological lineages exist in a nested or multigenerational state, however, testing this hypothesis is a difficult endeaver (O' Malley 2016). Transposons offer a promising model in which to study nested lineages, as transposon lineages are nested within organismal lineages and identifying populations of transposons can be done with often pre-existing whole-genome sequence data. Phylogenies of transposons can be produced by comparing transposon copies, however, many families of transposons have high mutation rates and produced phylogenies may suffer from low resolution. By constructing networks of transposons based on  sequence similarity, and subsequntly analyzing resulting similarity networks, hypotheses about transposon lineages can be tested without attempting to resolve a potentially unresolvable phylogeny. 

## Methods 
**1. Graphs and the computational problem** Schneider et al. (2021) produced a large sequence similarity net of transposons in a variety of eukaryotes and their conclusions were primarily based on the presence of clusters in the network (don't know how their clusters were derived--I did not find a description in their methods). Although they calculated some simple network metrics (node degree and clustering coefficient, at least) these were only present in supplemental data files and were not mentioned in the paper or its appendix. As such, I was curious to see what, if any, meaningful conclusions could be made from calculating simple network metrics on a similarity network. I performed this sort of analysis on a small network representing cluster mc476, from the gSchneider's much larger sequence similarity network.

**2. Approaches** I calculated closeness centrality, degree distribution, edge_betweeness, average neighbor degree, and clustering coefficient for an unweighted network derived from Schneider's cluster mc476. I attempt to interpret plots of the distribution of these metrics, along with GraphSpace graphs that represent such metrics, in order to determine what biological relevancy each metric may have when applied to an unweighted sequence similarity network. 

**3. Datasets** The file, 'SSN_mc476_edgelist.csv,' was downloaded from the GitHub repo affiliated with Schneider et al. (2021) and includes transposons from the nematode *C. brenneri*. This network was chosen for the simple reason that it was the only network uploaded to the group's GitHub page that included the edges present in the network. Other network-related files included node names and various metrics calculated for each node, along with the transposon sequences that were analyzed to produce all of the networks, however, producing networks from these files would have required extensive computation on my behalf (likely a multiple-day long BLAST!) Regardless, in these similarity networks, nodes represent a transposon consensus sequence (derived from a set of highly similar transposon copies in a single genome). Edges are drawn between nodes if, when BLASTed again each other, the nodes' corresponding sequences produce a bitscore greater than a threshold (I believe 10e−30, though for this specific network the threshold was not explicitly stated). 

**4. Output Types** Two kinds of diagrams were produced: histograms plotting the distribution of a given metric and GraphSpace graphs with node color and node size corresponding to a given metric. 

## Results
![plot](https://github.com/febreezioman/transposon-net/blob/cca5150abc6cadbb8f8c448da12fa0211aef6804/figures/composite1.png)
![plot](https://github.com/febreezioman/transposon-net/blob/7e605397e1034ed7aa557bdb8aa6ee643f88660a/figures/composite2.png)
**Figure 1** A. Closeness centrality distribution including all nodes in cluster mc476. B. Average neighbor degree distribution for nodes in cluster mc476 C. Degree distribution for nodes in cluster mc476 E. Edge betweeness distribution including all edges in cluster mc476. F. Log-adjusted edge betweeness distribution including all nodes in mc476 (a log function was directly applied to edge-betweeness.)

There is not an apparent logarithmic-like relationship in fig 1A (frequency of various node degrees,) and as such it appears that cluster mc476 is not scale-free (a variant of fig 1A on a log scale can be seen in the 'figures' repo: './figures/dd_log_bin_1.png') There is no clear relationship in the frequencies of clustering coefficient bins (fig 1B), although it is interesting that for clustering coefficients between 0.4 and 0.6, the distribution is approximatley even. Both closeness centrality and average neighbor degree have approximatley similar distributions, centered on ~0.008 and ~23, respectively (fig 1C and 1D). The distribution of edge betweeness initially appear to possibly follow a logarithmic distribution (1E), but when edge betweeness is adjusted to a logarithmic scale--which would produce a linear relationship if the initial relationship was actually logarithmic--almost no pattern in the distribution of edge betweeness scores is present(1F).

![plot](https://github.com/febreezioman/transposon-net/blob/9664937cfcdde8139f35bb5239c4c8f4639dc86e/figures/composite3.png)
**Figure 2** All graphs represent cluster mc476 data with nodes as transposon consensus sequences and edges as similarity between node sequences greater than a threshold. Graphs were visualized using GraphSpace and the 'cola' layout. A. Node size corresponds to average neighbor degree. Green-er nodes have greater average neighbor degree and blue-er nodes lesser so. B. Node size corresponds to closeness centrality. Purple-er nodes have a greater closeness centrality and grey/green-er nodes lesser. C. Node size correspond to clustering coefficient. Purple-er nodes have a greater clustering coefficient and turquiose-er nodes lesser. 

Average neighbor degree does not appear to relate strongly to the degree of a given node (fig 2A). For example, the node 4671 has a high average neighbor degree (it's single edge is connected to promiscuous 12934) while node 5223 also has a single edge, but it is connected to node 5218, which only has a degree of 3. Closeness centrality is intuitive--nodes more central to the graph, for lack of a better word--but also generally literally with the cola layout, and usually with higher degrees, achieve higher closeness centrality values (fig 2B). Clustering coefficient, again quite intuitively, highlights nodes that appear more clustered, generally nodes with higher degree that, under the cola layout (and since this network appears to only have one tight cluster), are central to the graph. 

## Discussion 
A transposon that is highly mutated most likely will produce fewer copies than it would otherwise, and as such will have less neighbors both because it has less offspring and because it is not very similar to many other nodes in the network. Considering the degree of a node and the average neighbor degree of a node could be particularly informative in this regard, but it is difficult to determine exactly to what extent these metrics will be informative given the information in the unweighted mc476 network. Still, I propose a hypothesis: high node degrees and average neighbor degrees positvely correlate with high fitness transposon copies that, in turn, on average have low mutation rates between copies. Although this hypothesis is difficult to test with an unweighted network, using BLAST bitscore to weight the edges (as is given in the edge file) and producing modified 'degree' and 'average neighbor degree' metrics, where degree is simply the sum of all edge weights for edges connected to a node, offer a clear path to test the this relationship. Additionally, highly fit transposons could produce tight clusters (measured by clustering coefficient) with low edge betweeness (modified for weighted edges): eg. another testable hypothesis. Additionally, the node labels on this graph were not particularly informative (though possibly would be so if I attempted to hunt to the orgins of each *C brenneri* consensus sequence), but if nodes could be more informatively labeled, then correlating network metrics to biologically relevant events could be directly evaluted. Should the nodes additionally come from species with a highly resolved phylogeny, and the aforementioned metrics are promising with regards to empirically testing the dyamics of nested lineages. 

## References
O’Malley, M.A. (2016). Reproduction Expanded: Multigenerational and Multilineal Units of Evolution. Philosophy of Science, 83, 835–847.

Schneider, L., Guo, Y., Birch, D. & Sarkies, P. (2021). Network‐based visualisation reveals new insights into transposable element diversity. Mol Syst Biol, 17, e9600.


